api/desecapi/management/commands/check-slaves.py

```diff
@@ -0,0 +1,80 @@
+from datetime import timedelta
+from socket import gethostbyname
+from time import sleep
+
+from django.conf import settings
+from django.core.mail import get_connection, mail_admins
+from django.core.management import BaseCommand
+from django.utils import timezone
+import dns.message, dns.query, dns.rdatatype
+
+from desecapi import pdns
+from desecapi.models import Domain
+
+
+def query_serial(zone, server):
+    query = dns.message.make_query(zone, 'SOA')
+    response = dns.query.tcp(query, server)
+
+    for rrset in response.answer:
+        if rrset.rdtype == dns.rdatatype.SOA:
+            return int(rrset[0].serial)
+    return None
+
+
+def get_serials():
+    return {zone['name']: zone['edited_serial'] for zone in pdns._pdns_get(pdns.NSMASTER, '/zones').json()}
+
+
+class Command(BaseCommand):
+    help = 'Check slaves for consistency with nsmaster.'
+
+    def add_arguments(self, parser):
+        parser.add_argument('domain-name', nargs='*',
+                            help='Domain name to check. If omitted, will check all recently published domains.')
+        parser.add_argument('--delay', type=int, default=60, help='Delay SOA checks to allow pending AXFRs to finish.')
+        parser.add_argument('--window', type=int, default=settings.WATCHDOG_WINDOW_SEC,
+                            help='Check domains that were published no longer than this many seconds ago.')
+
+    def handle(self, *args, **options):
+        threshold = timezone.now() - timedelta(seconds=options['window'])
+        recent_domain_names = Domain.objects.filter(published__gt=threshold).values_list('name', flat=True)
+        serials = {zone: serial for zone, serial in get_serials().items() if zone.rstrip('.') in recent_domain_names}
+
+        if options['domain-name']:
+            serials = {zone: serial for zone, serial in serials.items() if zone.rstrip('.') in options['domain-name']}
+
+        print('Sleeping for {} seconds before checking {} domains ...'.format(options['delay'], len(serials)))
+        sleep(options['delay'])
+
+        outdated_count = 0
+        outdated_slaves = set()
+        servers = {gethostbyname(server): server for server in settings.WATCHDOG_SLAVES}
+        output = []
+        for zone, local_serial in serials.items():
+            outdated = {}
```

Comment:
consider moving the checking logic for one zone into a separate method

Reply:
fixed in 0daa221da566747d7ca40e402ba5b76e93436827